From 2dd7aa9496625f43f43e5a4b84288e8b26b02277 Mon Sep 17 00:00:00 2001
From: Renjith Ravindran <Renjith.RavindranKannath@amd.com>
Date: Tue, 27 Feb 2024 19:19:40 +0000
Subject: [PATCH] For avx build,the start address of values_ buffer in
 KernelParameters is not correct as it is computed based on 16-byte alignment.

---
 clr/rocclr/platform/kernel.hpp | 6 +++---
 clr/rocclr/utils/flags.hpp     | 2 +-
 clr/rocclr/utils/macros.hpp    | 8 ++++++++
 3 files changed, 12 insertions(+), 4 deletions(-)

diff --git a/clr/rocclr/platform/kernel.hpp b/clr/rocclr/platform/kernel.hpp
index 8cb3b7f..435bc23 100644
--- a/clr/rocclr/platform/kernel.hpp
+++ b/clr/rocclr/platform/kernel.hpp
@@ -159,7 +159,7 @@ class KernelParameters : protected HeapObject {
         deviceKernelArgs_(false) {
     totalSize_ = signature.paramsSize() + (signature.numMemories() +
         signature.numSamplers() + signature.numQueues()) * sizeof(void*);
-    values_ = reinterpret_cast<address>(this) + alignUp(sizeof(KernelParameters), 16);
+    values_ = reinterpret_cast<address>(this) + alignUp(sizeof(KernelParameters), PARAMETERS_MIN_ALIGNMENT);
     memoryObjOffset_ = signature_.paramsSize();
     memoryObjects_ = reinterpret_cast<amd::Memory**>(values_ + memoryObjOffset_);
     samplerObjOffset_ = memoryObjOffset_ + signature_.numMemories() * sizeof(amd::Memory*);
@@ -183,7 +183,7 @@ class KernelParameters : protected HeapObject {
         execNewVcop_(rhs.execNewVcop_),
         execPfpaVcop_(rhs.execPfpaVcop_),
         deviceKernelArgs_(false) {
-    values_ = reinterpret_cast<address>(this) + alignUp(sizeof(KernelParameters), 16);
+    values_ = reinterpret_cast<address>(this) + alignUp(sizeof(KernelParameters),PARAMETERS_MIN_ALIGNMENT);
     memoryObjOffset_ = signature_.paramsSize();
     memoryObjects_ = reinterpret_cast<amd::Memory**>(values_ + memoryObjOffset_);
     samplerObjOffset_ = memoryObjOffset_ + signature_.numMemories() * sizeof(amd::Memory*);
@@ -220,7 +220,7 @@ class KernelParameters : protected HeapObject {
   //! Allocate memory for this instance as well as the required storage for
   //  the values_, defined_, and rawPointer_ arrays.
   void* operator new(size_t size, const KernelSignature& signature) {
-    size_t requiredSize = alignUp(size, 16) + signature.paramsSize() +
+    size_t requiredSize = alignUp(size, PARAMETERS_MIN_ALIGNMENT) + signature.paramsSize() +
       (signature.numMemories() + signature.numSamplers() + signature.numQueues()) *
        sizeof(void*);
     return AlignedMemory::allocate(requiredSize, PARAMETERS_MIN_ALIGNMENT);
diff --git a/clr/rocclr/utils/flags.hpp b/clr/rocclr/utils/flags.hpp
index 2c7f698..c273059 100644
--- a/clr/rocclr/utils/flags.hpp
+++ b/clr/rocclr/utils/flags.hpp
@@ -40,7 +40,7 @@ debug(size_t, CPU_MEMORY_GUARD_PAGE_SIZE, 64,                                 \
         "Size in KB of CPU memory guard page")                                \
 debug(size_t, CPU_MEMORY_ALIGNMENT_SIZE, 256,                                 \
         "Size in bytes for the default alignment for guarded memory on CPU")  \
-debug(size_t, PARAMETERS_MIN_ALIGNMENT, 16,                                   \
+debug(size_t, PARAMETERS_MIN_ALIGNMENT, NATIVE_ALIGNMENT_SIZE,                \
         "Minimum alignment required for the abstract parameters stack")       \
 debug(size_t, MEMOBJ_BASE_ADDR_ALIGN, 4*Ki,                                   \
         "Alignment of the base address of any allocate memory object")        \
diff --git a/clr/rocclr/utils/macros.hpp b/clr/rocclr/utils/macros.hpp
index aaf7e7b..8b65a2f 100644
--- a/clr/rocclr/utils/macros.hpp
+++ b/clr/rocclr/utils/macros.hpp
@@ -126,6 +126,14 @@
 #define IS_WINDOWS false
 #endif
 
+#if defined(__AVX512F__)
+#define NATIVE_ALIGNMENT_SIZE 64
+#elif defined(__AVX__)
+#define NATIVE_ALIGNMENT_SIZE 32
+#else
+#define NATIVE_ALIGNMENT_SIZE 16
+#endif
+
 #define IF_LEFT_true(x) x
 #define IF_LEFT_false(x)
 #define IF_RIGHT_true(x)
-- 
2.39.3

